README.md
# branglecrypt

Erlang / C bcrypt hashing library.

## Features

* Branglecrypt is lightweight, secure and easy to use.
* None of the NIFs run for a long time.

## Usage

There are functions to generate a salt `bcrypt:gen_salt`
and then use that salt to hash a password `bcrypt:hashpw`, but there are
also the following three convenience functions (with examples):

* hashpwsalt -- generate a salt and then use that salt to hash a password

    Hash = bcrypt:hashpwsalt("difficult2guess").

* checkpw -- check the password against the stored hash

    bcrypt:checkpw("difficult2guess", Stored_hash).

* dummy_checkpw -- run a dummy check that always returns false

    bcrypt:dummy_checkpw().

## License

BSD. For more details, view the `LICENSE` file.


CHANGELOG.md
# Changelog

## 1.0.0

* Changes
  * Removed random password generator and strong password check (to keep it lightweight).
  * Edited C code and deleted unused functions.

## 0.3.0 (2015-5-23)

* Changes
  * Added password generator and strong password check.

## 0.2.0 (2015-4-19)

* Enhancements
  * Made NIF more scheduler-friendly (it does not block for a long time).


test/bcrypt_tests.erl
-module(bcrypt_tests).
-include_lib("eunit/include/eunit.hrl").
-include("test_vectors.hrl").

openwall_test_() ->
    pairs(?PAIRS_1).

b_prefix_test_() ->
    pairs(?PAIRS_2).

long_pass_test_() ->
    pairs(?PAIRS_3).

consistency_test_() ->
    pairs(?PAIRS_4).

pairs(Pairs) ->
    [?_assert(Hash =:= bcrypt:hashpw(Pass, Salt)) ||
     {Pass, Salt, Hash} <- Pairs].

hash_check(Pass, Wrong1, Wrong2, Wrong3) ->
    H = bcrypt:hashpwsalt(Pass),
    [?_assert(bcrypt:checkpw(Pass, H) =:= true),
    ?_assert(bcrypt:checkpw(Wrong1, H) =:= false),
    ?_assert(bcrypt:checkpw(Wrong2, H) =:= false),
    ?_assert(bcrypt:checkpw(Wrong3, H) =:= false)].

hash_check_test_() ->
    hash_check("password", "passwor", "passwords", "pasword"),
    hash_check("hard2guess", "hardguess", "hard 2guess", "had2guess").

hash_check_extended_test_() ->
    hash_check("aáåäeéêëoôö", "aáåäeéêëoö", "aáåeéêëoôö", "aáå äeéêëoôö"),
    hash_check("aáåä eéêëoôö", "aáåä eéê ëoö", "a áåeé êëoôö", "aáå äeéêëoôö").

hash_check_nonascii_test_() ->
    hash_check("Сколько лет, сколько зим", "Сколько лет,сколько зим",
    "Сколько лет сколько зим", "Сколько лет, сколько"),
    hash_check("สวัสดีครับ", "สวัดีครับ", "สวัสสดีครับ", "วัสดีครับ").

hash_check_mixedchars_test_() ->
    hash_check("Я❤três☕ où☔", "Я❤tres☕ où☔", "Я❤três☕où☔", "Я❤três où☔").

dummy_check_test_() ->
    ?_assert(bcrypt:dummy_checkpw() =:= false).

salt_log_num_test_() ->
    [?_assert(lists:prefix("$2b$08$", bcrypt:gen_salt(8)) =:= true),
    ?_assert(lists:prefix("$2b$20$", bcrypt:gen_salt(20)) =:= true)].

salt_length_test_() ->
    [?_assert(length(bcrypt:gen_salt()) == 29),
    ?_assert(length(bcrypt:gen_salt(8)) == 29),
    ?_assert(length(bcrypt:gen_salt(20)) == 29),
    ?_assert(length(bcrypt:gen_salt("wrong input but still works")) == 29)].

salt_wrong_input_test_() ->
    [?_assert(lists:prefix("$2b$12$", bcrypt:gen_salt(3)) =:= true),
    ?_assert(lists:prefix("$2b$12$", bcrypt:gen_salt(32)) =:= true),
    ?_assert(lists:prefix("$2b$12$", bcrypt:gen_salt(["wrong type"])) =:= true)].

hash_wrong_input_test_() ->
    ?_assertError(badarg, bcrypt:hashpw("U*U", "$2a$05$CCCCCCCCCCCCCCCCCCC.")).


test/password_tests.erl
-module(password_tests).
-include_lib("eunit/include/eunit.hrl").

strong_password_test_() ->
    [?_assert(password:strong_password("hfksh6jsdf#") =:= true),
    ?_assert(password:strong_password("8auyk kjkjh") =:= true),
    ?_assert(password:strong_password("ty3uhi@ksd") =:= true)].

weaker_pass_nodigit_test_() ->
    Message = "The password should contain at least one digit and one punctuation character.",
    [?_assert(password:strong_password("hfkshjsdf#") =:= Message),
    ?_assert(password:strong_password("auyk kjkjh") =:= Message),
    ?_assert(password:strong_password("tyuhi@ksd") =:= Message)].

weaker_pass_nopunc_test_() ->
    Message = "The password should contain at least one digit and one punctuation character.",
    [?_assert(password:strong_password("hfksh6jsdf") =:= Message),
    ?_assert(password:strong_password("8auykkjkjh") =:= Message),
    ?_assert(password:strong_password("ty3uhiksd") =:= Message)].

generate_strong_test_() ->
    ?_assert(password:strong_password(password:gen_password()) =:= true).


src/bcrypt.erl
%%% @doc Erlang NIF implementation of OpenBSD Bcrypt hashing scheme.
%%%      Bcrypt is a key derivation function for passwords designed by Niels Provos
%%%      and David Mazières. Bcrypt is an adaptive hashing function, which means that
%%%      it can be configured to remain slow and resistant to brute-force attacks
%%%      even as computational power increases.
%%%      This bcrypt implementation is based on the latest OpenBSD version, which
%%%      fixed a small issue that affected some passwords longer than 72 characters.

-module(bcrypt).
-export([start/0, stop/0]).
-export([gen_salt/0, gen_salt/1, hashpw/2, checkpw/2, hashpwsalt/1, dummy_checkpw/0]).
-on_load(init/0).

-define(LOGR, 12).

start() ->
    application:start(bcrypt).
stop() ->
    application:stop(bcrypt).

%% @doc Initialize bcrypt NIF (the functions encode_salt and hashpw).
init() ->
    SoName = filename:join(case code:priv_dir(?MODULE) of
                               {error, bad_name} ->
                                   filename:join(
                                     [filename:dirname(
                                        code:which(?MODULE)),"..","priv"]);
                               Dir ->
                                   Dir
                           end, atom_to_list(?MODULE) ++ "_nif"),
    erlang:load_nif(SoName, 0).

%% @doc Initialize the P-box and S-box tables with the digits of Pi,
%%      and then start the key expansion process.
bf_init(_Key, _Key_len, _Salt) ->
    erlang:nif_error({nif_not_loaded, module, ?MODULE, line, ?LINE}).

%% @doc The main key expansion function. This function is called
%%      2^log_rounds times.
bf_expand(_State, _Key, _Key_len, _Salt) ->
    erlang:nif_error({nif_not_loaded, module, ?MODULE, line, ?LINE}).

%% @doc Encrypt and return the hash.
bf_encrypt(_State) ->
    erlang:nif_error({nif_not_loaded, module, ?MODULE, line, ?LINE}).

%% @doc Generate a salt for use with the hashpw function.
%%      The log_rounds parameter determines the computational complexity
%%      of the hashing. Its default is 12, the minimum is 4, and the maximum
%%      is 31.
gen_salt() ->
    gen_salt(?LOGR).

gen_salt(LogRounds) when is_integer(LogRounds), LogRounds > 4, LogRounds < 32 ->
    fmt_salt(binary:bin_to_list(crypto:strong_rand_bytes(16)), zero_str(LogRounds));
gen_salt(_LogR) ->
    gen_salt(?LOGR).

%% @doc Hash the password using bcrypt.
hashpw(Password, Salt) ->
    {Salt1, _} = lists:split(29, Salt),
    [Prefix, LogRounds, Salt2] = string:tokens(Salt1, "$"),
    Hash = bcrypt(filter_chars(Password, []), Salt2, Prefix, LogRounds),
    fmt_hash(Hash, Salt2, Prefix, LogRounds).

filter_chars([], Acc) -> Acc;
filter_chars([H|T], Acc) when H > 255 ->
    Chars = binary:bin_to_list(unicode:characters_to_binary([H])),
    filter_chars(T, lists:append(Acc, Chars));
filter_chars([H|T], Acc) ->
    filter_chars(T, lists:append(Acc, [H])).

bcrypt(Key, Salt, Prefix, LogRounds) ->
    Key_len = case Prefix of
        "2b" -> min(73, length(Key) + 1);
        _ -> length(Key) + 1
    end,
    {Salt1, Rounds} = prepare_keys(Salt, string:to_integer(LogRounds)),
    State = bf_init(Key, Key_len, Salt1),
    bf_encrypt(expand_keys(State, Key, Key_len, Salt1, Rounds)).

prepare_keys(Salt, {LogRounds, _}) when LogRounds > 4 andalso LogRounds < 32 ->
    {bbase64:decode(Salt), 1 bsl LogRounds};
prepare_keys(_, _) ->
    erlang:error({badarg}).

expand_keys(State, _Key, _Key_len, _Salt, 0) ->
    State;
expand_keys(State, Key, Key_len, Salt, Rounds) ->
    expand_keys(bf_expand(State, Key, Key_len, Salt),
                Key, Key_len, Salt, Rounds - 1).

zero_str(LogRounds) ->
    if LogRounds < 10 -> lists:concat(["0", LogRounds]);
       true -> LogRounds
    end.

fmt_salt(Salt, LogRounds) ->
    lists:concat(["$2b$", LogRounds, "$", bbase64:encode(Salt)]).

fmt_hash(Hash, Salt, Prefix, LogRounds) ->
    lists:concat(["$", Prefix, "$", LogRounds, "$", Salt, bbase64:encode(Hash)]).

%% @doc Convenience function that randomly generates a salt,
%%      and then hashes the password with that salt.
hashpwsalt(Password) ->
    hashpw(Password, gen_salt(?LOGR)).

%% @doc Check the password against the stored hash.
%%      The password and stored hash are compared in constant time
%%      to avoid timing attacks.
checkpw(Plaintext, Stored_hash) ->
    Hash = hashpw(Plaintext, Stored_hash),
    secure_check(Hash, Stored_hash).

%% @doc Perform a dummy check for a user that does not exist.
%%      This always returns false.
%%      The reason for implementing this check is in order to make
%%      user enumeration via timing attacks more difficult.
dummy_checkpw() ->
    hashpwsalt("password"),
    false.

secure_check(Hash, Stored) when is_list(Hash) and is_list(Stored) ->
    case length(Hash) == length(Stored) of
        true -> secure_check(Hash, Stored, 0);
        false -> false
    end;
secure_check(_Hash, _Stored) -> false.

secure_check([H|RestH], [S|RestS], Result) ->
    secure_check(RestH, RestS, (H bxor S) bor Result);
secure_check([], [], Result) ->
    Result == 0.


src/bbase64.erl
%%% @doc Adapted base64 implementation for bcrypt.

-module(bbase64).
-export([encode/1, decode/1]).

%% One-based decode map.
-define(DECODE_MAP,
	{bad,bad,bad,bad,bad,bad,bad,bad,ws,ws,bad,bad,ws,bad,bad, %1-15
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad, %16-31
	 ws,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,0,1, %32-47
	 54,55,56,57,58,59,60,61,62,63,bad,bad,bad,eq,bad,bad, %48-63
	 bad,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,
	 17,18,19,20,21,22,23,24,25,26,27,bad,bad,bad,bad,bad,
	 bad,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,
	 43,44,45,46,47,48,49,50,51,52,53,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad}).

encode(List) when is_list(List) ->
    encode_l(List).

encode_l([]) ->
    [];
encode_l([A]) ->
    [b64e(A bsr 2),
     b64e((A band 3) bsl 4)];
encode_l([A,B]) ->
    [b64e(A bsr 2),
     b64e(((A band 3) bsl 4) bor (B bsr 4)), 
     b64e((B band 15) bsl 2)];
encode_l([A,B,C|Ls]) ->
    BB = (A bsl 16) bor (B bsl 8) bor C,
    [b64e(BB bsr 18),
     b64e((BB bsr 12) band 63), 
     b64e((BB bsr 6) band 63),
     b64e(BB band 63) | encode_l(Ls)].

decode(List) when is_list(List) ->
    decode_l(List, []).

decode_l([], A) -> A;
decode_l([C1,C2], A) ->
    Bits2x6 = (b64d(C1) bsl 18) bor (b64d(C2) bsl 12),
    Octet1 = Bits2x6 bsr 16,
    lists:append(A, [Octet1]);
decode_l([C1,C2,C3], A) ->
    Bits3x6 = (b64d(C1) bsl 18) bor (b64d(C2) bsl 12)
	bor (b64d(C3) bsl 6),
    Octet1 = Bits3x6 bsr 16,
    Octet2 = (Bits3x6 bsr 8) band 16#ff,
    lists:append(A, [Octet1,Octet2]);
decode_l([C1,C2,C3,C4| Cs], A) ->
    Bits4x6 = (b64d(C1) bsl 18) bor (b64d(C2) bsl 12)
	bor (b64d(C3) bsl 6) bor b64d(C4),
    Octet1 = Bits4x6 bsr 16,
    Octet2 = (Bits4x6 bsr 8) band 16#ff,
    Octet3 = Bits4x6 band 16#ff,
    decode_l(Cs, lists:append(A, [Octet1,Octet2,Octet3])).

%% accessors 
b64e(X) ->
    element(X+1,
	    {$., $/, $A, $B, $C, $D, $E, $F, $G, $H, $I, $J, $K, $L,
             $M, $N, $O, $P, $Q, $R, $S, $T, $U, $V, $W, $X,
             $Y, $Z, $a, $b, $c, $d, $e, $f, $g, $h, $i, $j, $k, $l,
             $m, $n, $o, $p, $q, $r, $s, $t, $u, $v, $w, $x,
             $y, $z, $0, $1, $2, $3, $4, $5, $6, $7, $8, $9}).

b64d(X) ->
    b64d_ok(element(X, ?DECODE_MAP)).

b64d_ok(I) when is_integer(I) -> I;
b64d_ok(_) ->
    erlang:error({badarg}).


src/password.erl
-module(password).
-export([gen_password/0, gen_password/1, strong_password/1]).

-define(ALPHA, lists:seq($A, $Z) ++ lists:seq($a, $z)).
-define(DIGITS, "0123456789").
-define(RAND_PUNC, ",./!@#$%^&*();:?<>").
-define(PUNC, " ,./!@#$%^&*();:?<>").
-define(CHAR_MAP, list_to_tuple(?RAND_PUNC ++ ?ALPHA ++ ?DIGITS)).

-define(PASS_LEN, 12).
-define(MIN_PASS_LEN, 8).


%% @doc Function to generate a random password.
gen_password() ->
    rand_password(?PASS_LEN).
gen_password(Len) ->
    rand_password(Len).

rand_password(Len) ->
    case pass_check(rand_numbers(Len)) of
        false -> rand_password(Len);
        Code -> [get_char(X) || X <- Code]
    end.

rand_numbers(Len) ->
    [crypto:rand_uniform(1, 81) || _ <- lists:seq(1, Len)].

pass_check(Code) ->
    lists:any(fun(X) -> X < 19 end, Code) andalso
    lists:any(fun(X) -> X > 70 end, Code) andalso Code.

get_char(Val) ->
    element(Val, ?CHAR_MAP).

%% @doc Function that checks password strength.
strong_password(Password) ->
    case pass_length(length(Password), ?MIN_PASS_LEN) of
        true -> extra_chars(Password);
        Message -> Message
    end.

pass_length(Word_len, Min_len) when Word_len < Min_len ->
    "The password should be at least " ++ integer_to_list(?MIN_PASS_LEN) ++ " characters long.";
pass_length(_, _) ->
    true.

extra_chars(Word) ->
    case has_punc_digit(Word) of
        true -> true;
        _ -> "The password should contain at least one digit and one punctuation character."
    end.

has_punc_digit(Word) ->
    lists:any(fun(X) -> lists:member(X, ?DIGITS) end, Word) andalso
    lists:any(fun(X) -> lists:member(X, ?PUNC) end, Word).


c_src/bcrypt_nif.c
/*	$OpenBSD: bcrypt.c,v 1.52 2015/01/28 23:33:52 tedu Exp $	*/

/*
 * Copyright (c) 2014 Ted Unangst <tedu@openbsd.org>
 * Copyright (c) 1997 Niels Provos <provos@umich.edu>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
/* This password hashing algorithm was designed by David Mazieres
 * <dm@lcs.mit.edu> and works as follows:
 *
 * 1. state := InitState ()
 * 2. state := ExpandKey (state, salt, password)
 * 3. REPEAT rounds:
 *      state := ExpandKey (state, 0, password)
 *	state := ExpandKey (state, 0, salt)
 * 4. ctext := "OrpheanBeholderScryDoubt"
 * 5. REPEAT 64:
 * 	ctext := Encrypt_ECB (state, ctext);
 * 6. RETURN Concatenate (salt, ctext);
 *
 * This version is designed to not allow any NIF to run for too long,
 * and has been implemented by David Whitlock and Jason M Barnes.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "erl_nif.h"
#include "erl_blf.h"

#define BCRYPT_MAXSALT 16
#define BCRYPT_WORDS 6
#define	BCRYPT_HASHLEN 23

static void secure_bzero(void *, size_t);

static ERL_NIF_TERM bf_init(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
	ErlNifBinary state;
	char key[1024];
	char salt[1024];
	size_t key_len;
	unsigned long key_len_arg;
	uint8_t salt_len;

	if (argc != 3 || !enif_get_string(env, argv[0], key, sizeof(key), ERL_NIF_LATIN1) ||
			!enif_get_ulong(env, argv[1], &key_len_arg) ||
			!enif_get_string(env, argv[2], salt, sizeof(salt), ERL_NIF_LATIN1))
		return enif_make_badarg(env);
	key_len = (size_t) key_len_arg;
	salt_len = BCRYPT_MAXSALT;

	if (!enif_alloc_binary(sizeof(blf_ctx), &state))
		return enif_make_badarg(env);

	Blowfish_initstate((blf_ctx *) state.data);
	Blowfish_expandstate((blf_ctx *) state.data, (uint8_t *) salt,
			salt_len, (uint8_t *) key, key_len);

	return enif_make_binary(env, &state);
}

static ERL_NIF_TERM bf_expand(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
	ErlNifBinary state;
	char key[1024];
	char salt[1024];
	size_t key_len;
	unsigned long key_len_arg;
	uint8_t salt_len;

	if (argc != 4 || !enif_inspect_binary(env, argv[0], &state))
		return enif_make_badarg(env);
	if (!enif_get_string(env, argv[1], key, sizeof(key), ERL_NIF_LATIN1) ||
			!enif_get_ulong(env, argv[2], &key_len_arg) ||
			!enif_get_string(env, argv[3], salt, sizeof(salt), ERL_NIF_LATIN1))
		return enif_make_badarg(env);
	key_len = (size_t) key_len_arg;
	salt_len = BCRYPT_MAXSALT;

	Blowfish_expand0state((blf_ctx *) state.data, (uint8_t *) key, key_len);
	Blowfish_expand0state((blf_ctx *) state.data, (uint8_t *) salt, salt_len);

	return enif_make_binary(env, &state);
}

static ERL_NIF_TERM bf_encrypt(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
{
	ErlNifBinary state;
	uint32_t i, k, m;
	uint16_t j;
	uint8_t ciphertext[4 * BCRYPT_WORDS] = "OrpheanBeholderScryDoubt";
	uint32_t cdata[BCRYPT_WORDS];
	ERL_NIF_TERM encrypted[4 * BCRYPT_WORDS];

	/* Initialize our data from argv */
	if (argc != 1 || !enif_inspect_binary(env, argv[0], &state))
		return enif_make_badarg(env);

	/* This can be precomputed later */
	j = 0;
	for (i = 0; i < BCRYPT_WORDS; i++)
		cdata[i] = Blowfish_stream2word(ciphertext, 4 * BCRYPT_WORDS, &j);

	/* Now do the encryption */
	for (k = 0; k < 64; k++)
		blf_enc((blf_ctx *) state.data, cdata, BCRYPT_WORDS / 2);

	for (i = 0; i < BCRYPT_WORDS; i++) {
		ciphertext[4 * i + 3] = cdata[i] & 0xff;
		cdata[i] = cdata[i] >> 8;
		ciphertext[4 * i + 2] = cdata[i] & 0xff;
		cdata[i] = cdata[i] >> 8;
		ciphertext[4 * i + 1] = cdata[i] & 0xff;
		cdata[i] = cdata[i] >> 8;
		ciphertext[4 * i + 0] = cdata[i] & 0xff;
	}

	for (m = 0; m < BCRYPT_HASHLEN; m++) {
		encrypted[m] = enif_make_uint(env, ciphertext[m]);
	}
	secure_bzero(state.data, state.size);
	enif_release_binary(&state);
	secure_bzero(ciphertext, sizeof(ciphertext));
	secure_bzero(cdata, sizeof(cdata));
	return enif_make_list_from_array(env, encrypted, BCRYPT_HASHLEN);
}

/*
 * A typical memset() or bzero() call can be optimized away due to "dead store
 * elimination" by sufficiently intelligent compilers.  This is a problem for
 * the above bf_encrypt() function which tries to zero-out several temporary
 * buffers before returning.  If these calls get optimized away, then these
 * buffers might leave sensitive information behind.  There are currently no
 * standard, portable functions to handle this issue -- thus the
 * implementation below.
 *
 * This function cannot be optimized away by dead store elimination, but it
 * will be slower than a normal memset() or bzero() call.  Given that the
 * bcrypt algorithm is designed to consume a large amount of time, the change
 * will likely be negligible.
 */
static void
secure_bzero(void *buf, size_t len)
{
	if (buf == NULL || len == 0) {
		return;
	}

	volatile unsigned char *ptr = buf;
	while (len--) {
		*ptr++ = 0;
	}
}

static ErlNifFunc bcrypt_nif_funcs[] =
{
	{"bf_init", 3, bf_init},
	{"bf_expand", 4, bf_expand},
	{"bf_encrypt", 1, bf_encrypt}
};

ERL_NIF_INIT(bcrypt, bcrypt_nif_funcs, NULL, NULL, NULL, NULL)


src/tools.erl
%%% @doc Adapted base64 implementation for bcrypt.

-module(tools).
-export([encode64/1, decode64/1]).

%% One-based decode map.
-define(DECODE_MAP,
	{bad,bad,bad,bad,bad,bad,bad,bad,ws,ws,bad,bad,ws,bad,bad, %1-15
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad, %16-31
	 ws,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,0,1, %32-47
	 54,55,56,57,58,59,60,61,62,63,bad,bad,bad,eq,bad,bad, %48-63
	 bad,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,
	 17,18,19,20,21,22,23,24,25,26,27,bad,bad,bad,bad,bad,
	 bad,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,
	 43,44,45,46,47,48,49,50,51,52,53,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,
	 bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad,bad}).

encode64(List) when is_list(List) ->
    encode_l(List).

encode_l([]) ->
    [];
encode_l([A]) ->
    [b64e(A bsr 2),
     b64e((A band 3) bsl 4)];
encode_l([A,B]) ->
    [b64e(A bsr 2),
     b64e(((A band 3) bsl 4) bor (B bsr 4)), 
     b64e((B band 15) bsl 2)];
encode_l([A,B,C|Ls]) ->
    BB = (A bsl 16) bor (B bsl 8) bor C,
    [b64e(BB bsr 18),
     b64e((BB bsr 12) band 63), 
     b64e((BB bsr 6) band 63),
     b64e(BB band 63) | encode_l(Ls)].

decode64(List) when is_list(List) ->
    decode_l(List).

decode_l(List) ->
    decode(List, []).

decode([], A) -> A;
decode([C1,C2], A) ->
    Bits2x6 = (b64d(C1) bsl 18) bor (b64d(C2) bsl 12),
    Octet1 = Bits2x6 bsr 16,
    lists:append(A, [Octet1]);
decode([C1,C2,C3], A) ->
    Bits3x6 = (b64d(C1) bsl 18) bor (b64d(C2) bsl 12)
	bor (b64d(C3) bsl 6),
    Octet1 = Bits3x6 bsr 16,
    Octet2 = (Bits3x6 bsr 8) band 16#ff,
    lists:append(A, [Octet1,Octet2]);
decode([C1,C2,C3,C4| Cs], A) ->
    Bits4x6 = (b64d(C1) bsl 18) bor (b64d(C2) bsl 12)
	bor (b64d(C3) bsl 6) bor b64d(C4),
    Octet1 = Bits4x6 bsr 16,
    Octet2 = (Bits4x6 bsr 8) band 16#ff,
    Octet3 = Bits4x6 band 16#ff,
    decode(Cs, lists:append(A, [Octet1,Octet2,Octet3])).

%% accessors 
b64e(X) ->
    element(X+1,
	    {$., $/, $A, $B, $C, $D, $E, $F, $G, $H, $I, $J, $K, $L,
             $M, $N, $O, $P, $Q, $R, $S, $T, $U, $V, $W, $X,
             $Y, $Z, $a, $b, $c, $d, $e, $f, $g, $h, $i, $j, $k, $l,
             $m, $n, $o, $p, $q, $r, $s, $t, $u, $v, $w, $x,
             $y, $z, $0, $1, $2, $3, $4, $5, $6, $7, $8, $9}).

b64d(X) ->
    b64d_ok(element(X, ?DECODE_MAP)).

b64d_ok(I) when is_integer(I) -> I;
b64d_ok(_) ->
    erlang:error({badarg}).
